### PR TITLE
proxy: Publish network policy changes to L7 proxies via the xDS protocol

### PR DESCRIPTION
**Pull request by rlenglet**
Remove the filter name from xDS NetworkPolicy resource.

Convert L4 policies into NetworkPolicy resources and publish them to L7 proxies using the NPDS protocol. The conversion enforces a canonical representation of network policies (all lists are sorted, etc.) to allow identifying when a policy hasn't changed and minimize the volume of xDS gRPC messages exchanged with L7 proxies.

Define ResourceMutator.Clear to delete all resources of a given type.

Publish consumable policy updates to proxies through xDS. Use FromRequires clauses to refine the set of allowed remote identities for each proxy rule.

Fixes: https://github.com/cilium/cilium/issues/2596

Signed-off-by: Romain Lenglet <romain@covalent.io>

### CODE REVIEW

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
TODO: Remove `NetworkPolicy` resources when local identities become unused. This should be done by integrating with the kvstore cache, cf. @tgraf:

> The right model is to look at the kvstore cache and count the slave keys
> That is 100% reliable and accounts for expired keys when agents crashed and never came back up
> There is a `localKeys` table which maintains a refcnt per identity in use locally

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Ginkgo tests (unit tests) fail trying to write to /var/run/cilium/bootstrap.pb, for which root access would likely be needed. This could be solved by using the local test directory instead. However, when Ginkgo runs the unit tests it has not built the binaries yet, so the `cilium-envoy` binary does not exist, so Envoy cannot be run in a unit test. It is possible that your test does not depend on the binary running, and if that is the case then the generation of the bootstrap file should be suppressed altogether.

**Comment by jrajahalme** [comment]
Bash tests fail on the same permissions issue, even though they are run as root.

**Comment by rlenglet** [comment]
I've updated the endpoint code to not add/remove proxy redirects when the daemon is in dry run mode. Should take care of the test issue.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
The tests passed with the commit that disables redirects when the daemon is in dry-run mode:
https://jenkins.cilium.io/job/Cilium-PR-Bash-Tests-All/769/
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/772/
Rebasing...

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Another round of successful tests:
https://jenkins.cilium.io/job/Cilium-PR-Bash-Tests-All/778/flowGraphTable/
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/781/flowGraphTable/

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Another round of successful tests:
https://jenkins.cilium.io/job/Cilium-PR-Bash-Tests-All/781/
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/784/
Rebasing...

**Comment by rlenglet** [comment]
test-me-please

**Review by jrajahalme — APPROVED** [review]
Good work, Love the comprehensive tests!

File: pkg/envoy/server.go [hunk 1]
```diff
@@ -401,3 +405,108 @@ func createBootstrap(filePath string, name, cluster, version string, xdsSock, en
...
+		r.L7Rules = &cilium.PortNetworkPolicyRule_HttpRules{
+			HttpRules: &cilium.HttpNetworkPolicyRules{
+				HttpRules: httpRules,
+			},
+		}
+	default:
+		// No L7 parser means nothing for an L7 proxy to do. Ignore the rule.
+		return nil
+	}
+	// TODO: Support Kafka.
```
Comment (ianvernon) [inline pkg/envoy/server.go:440]: Can you add an issue number in the form GH-XXXX to this TODO?

File: pkg/envoy/server.go [hunk 2]
```diff
@@ -401,3 +405,108 @@ func createBootstrap(filePath string, name, cluster, version string, xdsSock, en
...
+			httpRules = append(httpRules, &cilium.HttpNetworkPolicyRule{Headers: headers})
+		}
+		SortHTTPNetworkPolicyRules(httpRules)
+		r.L7Rules = &cilium.PortNetworkPolicyRule_HttpRules{
+			HttpRules: &cilium.HttpNetworkPolicyRules{
+				HttpRules: httpRules,
+			},
+		}
+	default:
+		// No L7 parser means nothing for an L7 proxy to do. Ignore the rule.
```
Comment (ianvernon) [inline pkg/envoy/server.go:437]: If this check is done at the start of this function, it can exit much earlier - not very important, though.

**Review by ianvernon — APPROVED** [review]
LGTM - nothing blocking.

**Review by tgraf — APPROVED** [review]

**Pull request merged**